chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,177 @@
+Then the other namenodes from the zookeeper quorum then try to create an ephermal node, and the first one to do so aquires the lock, and becomes the Active namenode, only one Node can own the lock at a time so it guarantees that only one active namenode exists at a time.
+
+in the case of a split brain, the journal nodes will allow only one name node to have the role to write them at a time which removes the possible risk of data loss during a split brain.
+
+because the jouurnal nodes allow only one name node to have the role of writing to them at a time, in the case of the failover, the newly elected active name node only can write to the journal nodes which means that the other namenode won't write anything or in other words, won't be active as it dissallows it to be active, which in turn, allows for the new namenode to continue operations safely.
+
+and in the case of failover, the new Name node will read from the journal nodes and sync the FSImage so that no data would be lost.
+
+If HA is not turned on (default) there is no standby namenode and instead there is a secondary namenode which all it does is read the edit logs and build the FSImage so that if the Namenode falls, it won't have to rebuild the FSimage, which can be a heavy task and cause it to fail again or increase the downtime.
+
+In the case of name node failure, the cluster becomes unavailable up until the namenode restarts.
+
+
 7. **Q:** What is the role of the NameNode in HDFS?
+7. **A:** The role of the NameNode is to save metadata about the file and to manage where they are saved, in addition to managing and saving the FS structure and also replicate / delete replications as needed.
+the metadata is stored in memory and also in a persistant FSImage file, in addition to a local editlog which includes the logs of all the eddits that have happened to the file system structure.
+the namenodes also monitors all the data nodes and knows which are alive and which are down.
+
 8. **Q:** What is a DataNode in HDFS?
+8. **A:** A DataNode is the node that saves the data itself in the form of blocks with a max size that was configured beforehand in the hdfs-site.
+The datanode also performs the computations and analysis (such as running mapreduce jobs or spark processes, an example could be to detect fradulent activity over a long period of time) of the data on the machines which allows for less movement of data.
+the data nodes send a block report every 6 hours to the namenode which includes all the blocks stored in this data node, in addition to the name of the file, the path ("virtual path") and the block count.
+the data node stores metadata about the file as well, with the same name of the file (blk_{GeneratedNumber}) with the suffix ".meta" which includes the file name, the file path, the block count, the block id, permissions of the file, owner, updated at and other time metadata and file size.
+
 9. **Q:** How does HDFS achieve fault tolerance?
```

Comment:
I want you to describe the whole failover process. From the moment the active NN falls, to the moment the stand by becomes the active. Don't spare details here, it's the most important question. 
Please include:   
- Fencing
- ZKFC
- Quorum
- Journal Nodes
- NN startup process